### PR TITLE
CDAP-5450: emit metrics in workflow context along with program context

### PR DESCRIPTION
**Pull request by gsps1**
When we emit metrics from MR,Spark, if they run as a part of workflow, we emit another metric in workflow context.
JIRA : https://issues.cask.co/browse/CDAP-5450
Build: http://builds.cask.co/browse/CDAP-DUT3831-2

### CODE REVIEW

**Comment by gokulavasan** [comment]
@gsps1 Any unit tests for the change?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceContext.java [hunk 1]
```diff
@@ -402,4 +404,18 @@ private InputFormatProvider createInputFormatProvider(String datasetName,
...
+  private MetricsContext getWorkflowMetricsContext(@Nullable WorkflowProgramInfo workflowProgramInfo, Program program,
+                                                   MetricsCollectionService metricsCollectionService) {
+    if (workflowProgramInfo == null) {
+      return null;
+    }
+    Map<String, String> tags = Maps.newHashMap();
+    tags.put(Constants.Metrics.Tag.NAMESPACE, program.getNamespaceId());
+    tags.put(Constants.Metrics.Tag.APP, program.getApplicationId());
+    tags.put(Constants.Metrics.Tag.WORKFLOW, workflowProgramInfo.getName());
+    tags.put(Constants.Metrics.Tag.RUN_ID, workflowProgramInfo.getRunId().getId());
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceContext.java]: would it make sense to add the node name within the workflow, too?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceContext.java [hunk 2]
```diff
@@ -73,7 +74,8 @@
   private final MapReduceSpecification spec;
   private final LoggingContext loggingContext;
   private final WorkflowProgramInfo workflowProgramInfo;
-  private final Metrics userMetrics;
+  private final MetricsContext userMetricsContext;
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceContext.java]: should this be called programMetricsContext?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceTaskContext.java [hunk 3]
```diff
@@ -77,7 +78,8 @@
 
   private final MapReduceSpecification spec;
   private final WorkflowProgramInfo workflowProgramInfo;
-  private final Metrics userMetrics;
+  private final MetricsContext userMetricsContext;
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceTaskContext.java]: same here

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/AbstractSparkContext.java [hunk 4]
```diff
@@ -104,8 +108,9 @@ protected AbstractSparkContext(ApplicationSpecification applicationSpecification
     this.runId = runId;
     this.programClassLoader = programClassLoader;
     this.discoveryServiceClient = discoveryServiceClient;
-    this.userMetrics = new ProgramUserMetrics(metricsContext);
+    this.metricsCollectionService = metricsCollectionService;
     this.metricsContext = metricsContext;
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/AbstractSparkContext.java]: this also is the programMetricsContext?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/AbstractSparkContext.java [hunk 5]
```diff
@@ -257,7 +262,11 @@ public DiscoveryServiceClient getDiscoveryServiceClient() {
    * not serializable and shouldn't be exposed to user program directly.
    */
   public Metrics getUserMetrics() {
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/AbstractSparkContext.java]: why is this called getUserMetrics() if the resulting metrics instance should not be exposed to users?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceTaskContext.java [hunk 6]
```diff
@@ -216,9 +214,27 @@ private static MetricsContext getMetricCollector(Program program, String runId,
...
+                                                   MetricsCollectionService metricsCollectionService) {
+    if (workflowProgramInfo == null) {
+      return null;
+    }
+    Map<String, String> tags = Maps.newHashMap();
+    tags.put(Constants.Metrics.Tag.NAMESPACE, program.getNamespaceId());
+    tags.put(Constants.Metrics.Tag.APP, program.getApplicationId());
+    tags.put(Constants.Metrics.Tag.WORKFLOW, workflowProgramInfo.getName());
+    tags.put(Constants.Metrics.Tag.RUN_ID, workflowProgramInfo.getRunId().getId());
+    return metricsCollectionService.getContext(tags);
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceTaskContext.java]: add the node name?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/AbstractSparkContext.java [hunk 7]
```diff
@@ -330,6 +339,30 @@ protected static MetricsContext createMetricsContext(MetricsCollectionService se
...
+    if (workflowProgramInfo == null) {
+      return null;
+    }
+
+    Map<String, String> tags = Maps.newHashMap();
+    tags.put(Constants.Metrics.Tag.NAMESPACE, program.getNamespaceId());
+    tags.put(Constants.Metrics.Tag.APP, program.getApplicationId());
+    tags.put(Constants.Metrics.Tag.WORKFLOW, workflowProgramInfo.getName());
+    tags.put(Constants.Metrics.Tag.RUN_ID, workflowProgramInfo.getRunId().getId());
+    return metricsCollectionService.getContext(tags);
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/AbstractSparkContext.java]: add the workflow node name?

Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/AbstractSparkContext.java]: which is the same as the program name

**Comment by anew** [comment]
+1 for test cases. There should be existing test case that can be extended, right? 

Other than that, please see my comments, @gsps1

**Comment by gsps1** [comment]
Made a mistake as missed pulling develop with rebase tag, opening a separate PR.

**Pull request closed without merging**